chainstate/tx-verifier/src/transaction_verifier/tokens_accounting_undo_cache.rs

```diff
@@ -0,0 +1,144 @@
+}
+
+impl TokensAccountingBlockUndoCache {
+    pub fn new() -> Self {
+        Self {
+            data: BTreeMap::new(),
+        }
+    }
+
+    #[cfg(test)]
+    pub fn new_for_test(data: BTreeMap<TransactionSource, TokensAccountingBlockUndoEntry>) -> Self {
+        Self { data }
+    }
+
+    pub fn data(&self) -> &BTreeMap<TransactionSource, TokensAccountingBlockUndoEntry> {
+        &self.data
+    }
+
+    pub fn consume(self) -> BTreeMap<TransactionSource, TokensAccountingBlockUndoEntry> {
+        self.data
+    }
+
+    pub fn fetch_block_undo<F, E>(
+        &mut self,
+        tx_source: &TransactionSource,
+        fetcher_func: F,
+    ) -> Result<Option<&mut BlockUndo>, ConnectTransactionError>
+    where
+        F: Fn(Id<Block>) -> Result<Option<BlockUndo>, E>,
+        ConnectTransactionError: From<E>,
+    {
+        match self.data.entry(*tx_source) {
+            Entry::Occupied(entry) => Ok(Some(&mut entry.into_mut().undo)),
+            Entry::Vacant(entry) => match tx_source {
+                TransactionSource::Chain(block_id) => {
+                    let entry = fetcher_func(*block_id)?.map(|block_undo| {
+                        &mut entry
+                            .insert(TokensAccountingBlockUndoEntry {
+                                undo: block_undo,
+                                is_fresh: false,
+                            })
+                            .undo
+                    });
+                    Ok(entry)
+                }
+                TransactionSource::Mempool => Ok(None),
+            },
+        }
+    }
+
+    pub fn take_tx_undo<F, E>(
+        &mut self,
+        tx_source: &TransactionSource,
+        tx_id: &Id<Transaction>,
+        fetcher_func: F,
+    ) -> Result<Option<TxUndo>, ConnectTransactionError>
+    where
+        F: Fn(Id<Block>) -> Result<Option<BlockUndo>, E>,
+        ConnectTransactionError: From<E>,
+    {
+        Ok(self
+            .fetch_block_undo(tx_source, fetcher_func)?
+            .and_then(|entry| entry.take_tx_undo(tx_id)))
+    }
+
+    pub fn get_or_create_block_undo(&mut self, tx_source: &TransactionSource) -> &mut BlockUndo {
+        &mut self
+            .data
+            .entry(*tx_source)
+            .or_insert(TokensAccountingBlockUndoEntry {
+                is_fresh: true,
+                undo: Default::default(),
+            })
+            .undo
+    }
+
+    pub fn set_undo_data(
+        &mut self,
+        tx_source: TransactionSource,
+        new_undo: &BlockUndo,
+    ) -> Result<(), BlockUndoError> {
+        match self.data.entry(tx_source) {
+            Entry::Vacant(e) => {
+                e.insert(TokensAccountingBlockUndoEntry {
+                    undo: new_undo.clone(),
+                    is_fresh: true,
+                });
+            }
+            Entry::Occupied(mut e) => {
+                e.get_mut().undo.combine(new_undo.clone())?;
+            }
```

Comment:
Nice catch, I believe it's a bug, trying to replicate it in tests

Reply:
Actually now I think it is correct. `is_fresh` indicated whether it's available in parent. This code is called when we flush from one verifier to another. 
Let's consider a hierarchy: db <- verifier1 <- verifier2. If verifier1 inserted the data with is_fresh = true it will always be false for verifier2. And the flag in verifier1 should be preserver when verifier2 is flushed.
Now if verifier1 is with is_fresh = false it will also be false for verifier2. So we can never be in a situation of potential change from false to true.

Reply:
added some additional tests